# About The Dataset
This data is all about the sales data of mobile phones of various companies. Here we are predicting the price range of different mobile manufactured by various companies by finding out some relation between features of a mobile phone(eg:- RAM,Internal Memory etc)

In [11]:
import pandas as pd
import numpy as np
import tensorflow as tf

In [12]:
pd.options.mode.chained_assignment = None  # default='warn'

In [14]:
# Download training data
df_training = pd.read_csv("ADMN5015 CLASSIFIER TRAININGDATA_NIDIYA.csv")
df_training

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0               842     0          2.2         0   1       0           7   
1              1021     1          0.5         1   0       1          53   
2               563     1          0.5         1   2       1          41   
3               615     1          2.5         0   0       0          10   
4              1821     1          1.2         0  13       1          44   
...             ...   ...          ...       ...  ..     ...         ...   
1995            794     1          0.5         1   0       1           2   
1996           1965     1          2.6         1   0       0          39   
1997           1911     0          0.9         1   1       1          36   
1998           1512     0          0.9         0   4       1          46   
1999            510     1          2.0         1   5       1          45   

      m_dep  mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  \
0       0.6        188        2  ...         20       756  2549     9     7   
1       0.7        136        3  ...        905      1988  2631    17     3   
2       0.9        145        5  ...       1263      1716  2603    11     2   
3       0.8        131        6  ...       1216      1786  2769    16     8   
4       0.6        141        2  ...       1208      1212  1411     8     2   
...     ...        ...      ...  ...        ...       ...   ...   ...   ...   
1995    0.8        106        6  ...       1222      1890   668    13     4   
1996    0.2        187        4  ...        915      1965  2032    11    10   
1997    0.7        108        8  ...        868      1632  3057     9     1   
1998    0.1        145        5  ...        336       670   869    18    10   
1999    0.9        168        6  ...        483       754  3919    19     4   

      talk_time  three_g  touch_screen  wifi  price_range  
0            19        0             0     1            1  
1             7        1             1     0            2  
2             9        1             1     0            2  
3            11        1             0     0            2  
4            15        1             1     0            1  
...         ...      ...           ...   ...          ...  
1995         19        1             1     0            0  
1996         16        1             1     1            2  
1997          5        1             1     0            3  
1998         19        1             1     1            0  
1999          2        1             1     1            3  

[2000 rows x 21 columns]

In [15]:
# Make sure data types are correct (no objects)

df_training.dtypes

battery_power      int64
blue               int64
clock_speed      float64
dual_sim           int64
fc                 int64
four_g             int64
int_memory         int64
m_dep            float64
mobile_wt          int64
n_cores            int64
pc                 int64
px_height          int64
px_width           int64
ram                int64
sc_h               int64
sc_w               int64
talk_time          int64
three_g            int64
touch_screen       int64
wifi               int64
price_range        int64
dtype: object

Here the data type of every column is integer or float so we are good to go. 

In [10]:
# Pop label

target = df_training.pop('price_range')

In [17]:
# Define features

numeric_feature_names = ['battery_power', 'blue', 'clock_speed', 'dual_sim','fc','four_g','int_memory','m_dep','mobile_wt','n_cores','pc','px_height','px_width','ram','sc_h','sc_w','talk_time','three_g','touch_screen','wifi']
numeric_features = df_training[numeric_feature_names]
numeric_features.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  pc  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2   2         20       756  2549     9     7         19   
1        136        3   6        905      1988  2631    17     3          7   
2        145        5   6       1263      1716  2603    11     2          9   
3        131        6   9       1216      1786  2769    16     8         11   
4        141        2  14       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  
0        0             0     1  
1        1             1     0  
2        1             1     0  
3        1             0     0  
4        1             1     0

In [18]:
# Convert data to Tensorflow tensor

numeric_features = tf.convert_to_tensor(numeric_features)
numeric_features

<tf.Tensor: shape=(2000, 20), dtype=float64, numpy=
array([[8.420e+02, 0.000e+00, 2.200e+00, ..., 0.000e+00, 0.000e+00,
        1.000e+00],
       [1.021e+03, 1.000e+00, 5.000e-01, ..., 1.000e+00, 1.000e+00,
        0.000e+00],
       [5.630e+02, 1.000e+00, 5.000e-01, ..., 1.000e+00, 1.000e+00,
        0.000e+00],
       ...,
       [1.911e+03, 0.000e+00, 9.000e-01, ..., 1.000e+00, 1.000e+00,
        0.000e+00],
       [1.512e+03, 0.000e+00, 9.000e-01, ..., 1.000e+00, 1.000e+00,
        1.000e+00],
       [5.100e+02, 1.000e+00, 2.000e+00, ..., 1.000e+00, 1.000e+00,
        1.000e+00]])>

In [19]:
# Normalize data

normalizer = tf.keras.layers.Normalization(axis=-1)
normalizer.adapt(numeric_features)

In [21]:
# Define model

def get_basic_model():
  model = tf.keras.Sequential([
    normalizer,
    tf.keras.layers.Dense(24, activation='relu'),
    tf.keras.layers.Dense(24, activation='relu'),
    tf.keras.layers.Dense(4) # Possible labels 0, 1 ,2 or 3
  ])

  model.compile(optimizer='adam',
                loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                # loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
                metrics=['accuracy'])
  return model

In [26]:
# Train model

BATCH_SIZE = 2002

model = get_basic_model()
model.fit(numeric_features, target, epochs=700, batch_size=BATCH_SIZE)

Epoch 1/700
1/1 [==============================] - 1s 630ms/step - loss: 1.6027 - accuracy: 0.2280
Epoch 2/700
1/1 [==============================] - 0s 11ms/step - loss: 1.5904 - accuracy: 0.2290
Epoch 3/700
1/1 [==============================] - 0s 10ms/step - loss: 1.5785 - accuracy: 0.2270
Epoch 4/700
1/1 [==============================] - 0s 12ms/step - loss: 1.5670 - accuracy: 0.2310
Epoch 5/700
1/1 [==============================] - 0s 10ms/step - loss: 1.5558 - accuracy: 0.2330
Epoch 6/700
1/1 [==============================] - 0s 10ms/step - loss: 1.5451 - accuracy: 0.2345
Epoch 7/700
1/1 [==============================] - 0s 17ms/step - loss: 1.5347 - accuracy: 0.2330
Epoch 8/700
1/1 [==============================] - 0s 23ms/step - loss: 1.5247 - accuracy: 0.2370
Epoch 9/700
1/1 [==============================] - 0s 13ms/step - loss: 1.5151 - accuracy: 0.2365
Epoch 10/700
1/1 [==============================] - 0s 13ms/step - loss: 1.5057 - accuracy: 0.2390
Epoch 11/700
1/1 [

Here I put Batch size as 2002 and Epoch as 700. In the initial steps we can see that loss: 1.6027 - accuracy: 0.2280. When we kept on moving around Epoch = 500 it becomes loss: 0.0824 - accuracy: 0.9945, and when Epoch = 700, we will gwt something like this, loss: 0.0403 - accuracy: 1.0000. In this case the value of loss is much smaller than the intial one and we get an accuracy of 100% in the last one while it was 22% in the initial one.  



In [40]:
# Download test data

df_test = pd.read_csv("ADMN5015 CLASSIFIER TESTDATA_NIDIYA.csv")
df_test.head()

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0   1           1043     1          1.8         1  14       0           5   
1   2            841     1          0.5         1   4       1          61   
2   3           1807     1          2.8         0   1       0          27   
3   4           1546     0          0.5         1  18       1          25   
4   5           1434     0          1.4         0  11       1          49   

   m_dep  mobile_wt  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0    0.1        193  ...        226      1412  3476    12     7          2   
1    0.8        191  ...        746       857  3895     6     0          7   
2    0.9        186  ...       1270      1366  2396    17    10         10   
3    0.5         96  ...        295      1752  3893    10     0          7   
4    0.5        108  ...        749       810  1773    15     8          7   

   three_g  touch_screen  wifi  price_range  
0        0             1     0            1  
1        1             0     0            2  
2        0             1     1            2  
3        1             1     0            2  
4        1             0     1            1  

[5 rows x 22 columns]

In [41]:
# Pop label

test_target = df_test.pop('price_range')

In [42]:
# Define test features

test_numeric_feature_names = ['battery_power', 'blue', 'clock_speed', 'dual_sim','fc','four_g','int_memory','m_dep','mobile_wt','n_cores','pc','px_height','px_width','ram','sc_h','sc_w','talk_time','three_g','touch_screen','wifi']
test_numeric_features = df_test[test_numeric_feature_names]
test_numeric_features.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0           1043     1          1.8         1  14       0           5    0.1   
1            841     1          0.5         1   4       1          61    0.8   
2           1807     1          2.8         0   1       0          27    0.9   
3           1546     0          0.5         1  18       1          25    0.5   
4           1434     0          1.4         0  11       1          49    0.5   

   mobile_wt  n_cores  pc  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        193        3  16        226      1412  3476    12     7          2   
1        191        5  12        746       857  3895     6     0          7   
2        186        3   4       1270      1366  2396    17    10         10   
3         96        8  20        295      1752  3893    10     0          7   
4        108        6  18        749       810  1773    15     8          7   

   three_g  touch_screen  wifi  
0        0             1     0  
1        1             0     0  
2        0             1     1  
3        1             1     0  
4        1             0     1

In [43]:
# Convert test data to Tensorflow tensor

test_numeric_features = tf.convert_to_tensor(test_numeric_features)
test_numeric_features

<tf.Tensor: shape=(1000, 20), dtype=float64, numpy=
array([[1.043e+03, 1.000e+00, 1.800e+00, ..., 0.000e+00, 1.000e+00,
        0.000e+00],
       [8.410e+02, 1.000e+00, 5.000e-01, ..., 1.000e+00, 0.000e+00,
        0.000e+00],
       [1.807e+03, 1.000e+00, 2.800e+00, ..., 0.000e+00, 1.000e+00,
        1.000e+00],
       ...,
       [1.185e+03, 0.000e+00, 1.400e+00, ..., 1.000e+00, 0.000e+00,
        0.000e+00],
       [1.533e+03, 1.000e+00, 5.000e-01, ..., 0.000e+00, 1.000e+00,
        0.000e+00],
       [1.270e+03, 1.000e+00, 5.000e-01, ..., 1.000e+00, 0.000e+00,
        1.000e+00]])>

In [44]:
# Normalize data

normalizer = tf.keras.layers.Normalization(axis=-1)
normalizer.adapt(test_numeric_features)

In [45]:
# Evaluate test data based on training data

score = model.evaluate(test_numeric_features, test_target, verbose=1)
print(f'Test loss: {score[0]} / Test accuracy: {score[1]}')

32/32 [==============================] - 0s 1ms/step - loss: 14.1584 - accuracy: 0.2550
Test loss: 14.15837574005127 / Test accuracy: 0.2549999952316284


In [39]:
# Save model (optional)

 # model.save('price_model')

In [46]:
# Download cases to be predicted

df_predict = pd.read_csv("ADMN5015 CLASSIFIER PREDICTDATA_NIDIYA.csv")
df_predict

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0   1            623     1          0.5         1   5       1          29   
1   2            700     1          1.3         0   2       0          40   
2   3            995     0          2.6         1   4       0           7   
3   4           1325     0          1.3         1  10       1          33   
4   5           1982     1          2.6         1   7       1          43   

   m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0    0.9        134  ...  10        273      1814   717    12     5   
1    0.1        150  ...  11        612      1325  2537    15    10   
2    0.8        157  ...   6       1315      1804  1272    16     6   
3    0.9        165  ...  12       1653      1967   726    10     2   
4    0.3        113  ...   8       1409      1679  1791     9     2   

   talk_time  three_g  touch_screen  wifi  
0         13        1             0     0  
1          8        1             1     1  
2          8        0             1     1  
3         12        1             1     1  
4          8        1             1     0  

[5 rows x 21 columns]

In [47]:
# Make sure data types are correct (no objects)

df_predict.dtypes

id                 int64
battery_power      int64
blue               int64
clock_speed      float64
dual_sim           int64
fc                 int64
four_g             int64
int_memory         int64
m_dep            float64
mobile_wt          int64
n_cores            int64
pc                 int64
px_height          int64
px_width           int64
ram                int64
sc_h               int64
sc_w               int64
talk_time          int64
three_g            int64
touch_screen       int64
wifi               int64
dtype: object

Here the data type of every column is integer or float so we are good to go. 

In [48]:
# Convert data to Tensorflow tensor

predict_numeric_features = tf.convert_to_tensor(df_predict)
predict_numeric_features

<tf.Tensor: shape=(5, 21), dtype=float64, numpy=
array([[1.000e+00, 6.230e+02, 1.000e+00, 5.000e-01, 1.000e+00, 5.000e+00,
        1.000e+00, 2.900e+01, 9.000e-01, 1.340e+02, 3.000e+00, 1.000e+01,
        2.730e+02, 1.814e+03, 7.170e+02, 1.200e+01, 5.000e+00, 1.300e+01,
        1.000e+00, 0.000e+00, 0.000e+00],
       [2.000e+00, 7.000e+02, 1.000e+00, 1.300e+00, 0.000e+00, 2.000e+00,
        0.000e+00, 4.000e+01, 1.000e-01, 1.500e+02, 8.000e+00, 1.100e+01,
        6.120e+02, 1.325e+03, 2.537e+03, 1.500e+01, 1.000e+01, 8.000e+00,
        1.000e+00, 1.000e+00, 1.000e+00],
       [3.000e+00, 9.950e+02, 0.000e+00, 2.600e+00, 1.000e+00, 4.000e+00,
        0.000e+00, 7.000e+00, 8.000e-01, 1.570e+02, 1.000e+00, 6.000e+00,
        1.315e+03, 1.804e+03, 1.272e+03, 1.600e+01, 6.000e+00, 8.000e+00,
        0.000e+00, 1.000e+00, 1.000e+00],
       [4.000e+00, 1.325e+03, 0.000e+00, 1.300e+00, 1.000e+00, 1.000e+01,
        1.000e+00, 3.300e+01, 9.000e-01, 1.650e+02, 1.000e+00, 1.200e+01,
        1.6

In [49]:
# Normalize

normalizer = tf.keras.layers.Normalization(axis=-1)
normalizer.adapt(predict_numeric_features)

In [67]:
from tensorflow.python.ops.gen_math_ops import mod_eager_fallback
# Predict labels

class_names = ['0','1','2','3']

predictions = reconstructed_model.predict(predict_numeric_features)

# Create new columns in dataframe
df_predict['label'] = None
df_predict['certainty'] = None

for i, logits in enumerate(predictions):
  class_idx = tf.argmax(logits).numpy()
  p = tf.nn.softmax(logits)[class_idx]
  name = class_names[class_idx]
  print("Example {} prediction: {} ({:4.1f}%)".format(i, name, 100*p))

  # Save predictions to dataframe
  df_predict["label"].iloc[i] = name
  df_predict['certainty'].iloc[i] = format(p)


NameError: ignored

Benefits of Using Classifier Tool: 

1.Very useful in contemplating demand forecasts.

2.Planning workforce and customer churn analysis.

3.In-depth analysis of the competitors.

4.Forecasting external factors that can affect your workflow.

5.Fleet maintenance.

6.Identifying financial risks and modeling credit.